src/shared/sol-util-linux.c

```diff
@@ -0,0 +1,292 @@
+
+    return size;
+}
+
+void *
+sol_util_load_file_raw(const int fd, size_t *size)
+{
+    struct stat st;
+    int saved_errno;
+    char *tmp = NULL;
+    char *buffer = NULL;
+    size_t buffer_size = 0;
+    ssize_t ret;
+
+    if (fd < 0 || !size)
+        return NULL;
+
+    *size = 0;
+    if (fstat(fd, &st) >= 0 && st.st_size) {
+        buffer_size = st.st_size;
+        buffer = malloc(buffer_size);
+        if (!buffer)
+            goto err;
+
+        ret = _fill_buffer(fd, buffer, buffer_size, size);
+        if (ret <= 0)
+            goto end;
+    }
+
+    do {
+        buffer_size += CHUNK_SIZE;
+        tmp = realloc(buffer, buffer_size);
+        if (!tmp)
+            goto err;
+        buffer = tmp;
+
+        ret = _fill_buffer(fd, buffer + *size, CHUNK_SIZE, size);
+    } while (ret > 0);
+
+end:
+    if (ret < 0 || !*size)
+        goto err;
+
+    if (*size < buffer_size) {
+        tmp = realloc(buffer, *size);
+        if (tmp)
+            buffer = tmp;
+    }
+
+    return buffer;
+
+err:
+    saved_errno = errno;
+    free(buffer);
+    errno = saved_errno;
+    *size = 0;
+    return NULL;
+}
+
+char *
+sol_util_load_file_string(const char *filename, size_t *size)
+{
+    int fd = -1, saved_errno;
+    size_t read;
+    char *tmp, *data = NULL;
+
+    fd = open(filename, O_RDONLY | O_CLOEXEC);
+    if (fd < 0)
+        goto open_err;
+
+    data = sol_util_load_file_raw(fd, &read);
+    if (!data) {
+        data = strdup("");
+        read = 1;
+    } else if (data[read - 1] != '\0') {
+        tmp = realloc(data, read + 1);
+        if (!tmp)
+            goto err;
+        data = tmp;
+        data[read] = '\0';
+    }
+
+    close(fd);
+    if (size)
+        *size = read;
+    return data;
+
+err:
+    saved_errno = errno;
+    free(data);
+    close(fd);
+    errno = saved_errno;
+open_err:
+    if (size)
+        *size = 0;
+    return NULL;
+}
+
+static int
+get_progname(char *out, size_t size)
+{
+    char cwd[PATH_MAX] = { 0 };
+
+#ifdef HAVE_SYS_AUXV_H
+    char *execfn = NULL;
+#else
+    char execfn[PATH_MAX] = { 0 };
+#endif
+
+    if (getcwd(cwd, sizeof(cwd)) == NULL)
+        return -errno;
+
+#ifdef HAVE_SYS_AUXV_H
+    execfn = (char *)getauxval(AT_EXECFN);
+
+    if (!execfn)
+        return -errno;
+#else
+    if (readlink("/proc/self/exe", execfn, sizeof(execfn) - 1) < 0 || !execfn[0])
+        return -ENOSYS;
+#endif
+
+    if (execfn[0] == '/')
+        return snprintf(out, size, "%s", execfn);
+    else
+        return snprintf(out, size, "%s/%s", cwd, execfn);
+}
+
+int
+sol_util_get_rootdir(char *out, size_t size)
+{
+    char progname[PATH_MAX] = { 0 }, *substr, *prefix;
+    int r;
+
+    r = get_progname(progname, sizeof(progname));
+    if (r < 0 || r >= (int)sizeof(progname))
+        return r;
+
+    substr = strstr(progname, PREFIX);
+    if (!substr) {
+        return -1;
+    }
+
+    prefix = strndupa(progname, strlen(progname) - strlen(substr));
+    if (!prefix)
```

Comment:
It's undefined what happens when `alloca()` fails, so please don't check. Use `%.*s` in the snprintf() below.